backend/src/crd/controller/scheduledworkflow/controller.go

```diff
@@ -457,65 +465,107 @@ func (c *Controller) submitNextWorkflowIfNeeded(ctx context.Context, swf *util.S
 	}
-
-	workflow, err = c.submitNewWorkflowIfNotAlreadySubmitted(ctx, swf, nextScheduledEpoch, nowEpoch)
+	var workflowName string
+	submitted, workflowName, err = c.submitNewWorkflowIfNotAlreadySubmitted(ctx, swf, nextScheduledEpoch, nowEpoch)
 	if err != nil {
 		log.WithFields(log.Fields{
 			ScheduledWorkflow: swf.Name,
 		}).Errorf("Submitting workflow for ScheduledWorkflow (%v): transient error while submitting workflow: %v",
 			swf.Name, err)
 		// There was an error submitting a new workflow.
 		// We should attempt to handle the schedule again at a later time.
-		return nil, nextScheduledEpoch, err
+		return false, nextScheduledEpoch, err
 	}
 	log.WithFields(log.Fields{
 		ScheduledWorkflow: swf.Name,
-		Workflow:          workflow.Get().Name,
+		Workflow:          workflowName,
 	}).Infof("Submitting workflow for ScheduledWorkflow (%v): workflow (%v) successfully submitted (scheduled at: %v)",
-		swf.Name, workflow.Get().Name, commonutil.FormatTimeForLogging(nextScheduledEpoch))
-	return workflow, nextScheduledEpoch, nil
+		swf.Name, workflowName, commonutil.FormatTimeForLogging(nextScheduledEpoch))
+	return submitted, nextScheduledEpoch, nil
 }
 
 func (c *Controller) submitNewWorkflowIfNotAlreadySubmitted(
 	ctx context.Context,
 	swf *util.ScheduledWorkflow, nextScheduledEpoch int64, nowEpoch int64) (
-	*commonutil.Workflow, error) {
+	bool, string, error) {
 
 	workflowName := swf.NextResourceName()
 
 	// Try to fetch this workflow
 	// If it already exists, it means that it was already created in a previous iteration
 	// of this controller but that the controller failed to save this data.
-	foundWorkflow, isNotFoundError, err := c.workflowClient.Get(swf.Namespace,
+	_, isNotFoundError, err := c.workflowClient.Get(swf.Namespace,
 		workflowName)
 	if err == nil {
 		// The workflow was already created by a previous iteration of this controller.
 		// Nothing to do except returning the information needed by the controller to update
 		// the ScheduledWorkflow status.
-		return foundWorkflow, nil
+		return true, workflowName, nil
 	}
 
 	if !isNotFoundError {
 		// There was an error while attempting to retrieve the workflow
-		return nil, err
+		return false, "", err
 	}
 
-	// If the workflow is not found, we need to create it.
-	newWorkflow, err := swf.NewWorkflow(nextScheduledEpoch, nowEpoch)
-	createdWorkflow, err := c.workflowClient.Create(ctx, swf.Namespace, newWorkflow)
+	if swf.Spec.PipelineID != nil {
+
+		log.WithFields(log.Fields{
+			ScheduledWorkflow: swf.Name,
+			Workflow:          workflowName,
+		}).Infof("PipelineID used when submitting run",
+			swf.Name, workflowName, commonutil.FormatTimeForLogging(nextScheduledEpoch))
+
+		// The pipeline version is not allways set, optional.
+		resourceRefs := []*api.ResourceReference{&api.ResourceReference{
+			Key: &api.ResourceKey{
+				Type: api.ResourceType_PIPELINE,
+				Id:   *swf.Spec.PipelineVersionID,
+			},
+		}}
+		if swf.Spec.PipelineVersionID != nil {
```

Comment:
discussion: can you confirm KFP API behavior, I recall that it's valid to just specify the pipeline version ID, because we can query API to get corresponding pipeline.

Therefore, I think behavior here can be that:
* when version ID is specified, but not pipeline ID, we should still be able to process it.

What do you think? Are there challenges here?

Reply:
Makes sense!